Add classmethod get_entity to send Vehicle ID Req without TCP connect

HarshaLaxman opened the pull request:
#17

```
addr, resp = DoIPClient.get_entity()
```

Binds to dynamically assigned port (UDP_TEST_EQUIPMENT_REQUEST) and sends to UDP_DISCOVERY port. Send IP defaults to broadcast, with the option to send unicast by passing in a IP address (IPV4 only for now)

Created a staticmethod to create the UDP socket that is used by `await_vehicle_announcement` and `get_entity`. Since the latter calls the former after sending the request, we reuse the socket.

HarshaLaxman commented on doipclient/client.py:
```diff
@@ -288,6 +296,65 @@ def await_vehicle_announcement(
+        payload_data = message.pack()
+        payload_type = payload_message_to_type[type(message)]
+
+        data_bytes = struct.pack(
```
some code duplication from `send_doip` https://github.com/jacobschaer/python-doipclient/blob/1a1403eb7bb741139a5c32d8c0f9ff593fea2f02/doipclient/client.py#L415-L429

I can put it in a staticfunction if you want @jacobschaer ?

jacobschaer replied:
I see what you mean - that is somewhat inconvenient. Probably should move that out to a new function like you suggested, though leave the logger message out. I think it's ok for this new method to have its own unique logger message


```
@staticmethod
def pack_doip(protocol_version, payload_type, payload_data):
     data_bytes = struct.pack( 
	     "!BBHL", 
	     self._protocol_version, 
	     0xFF ^ self._protocol_version, 
	     payload_type, 
	     len(payload_data), 
	 ) 
	data_bytes += payload_data
	return data_bytes
```

Then in send_doip()... call it with

```
self.pack_doip(self._protocol_version, payload_type, payload_data)
logger.debug( 
     "Sending DoIP Message: Type: 0x{:X}, Payload Size: {}, Payload: {}".format( 
         payload_type, 
         len(payload_data), 
         " ".join(f"{byte:02X}" for byte in payload_data), 
     ) 
)
```

And in your new function maybe some new log message like "Sending DoIP Vehicle Information Request..."

HarshaLaxman replied:
Done!

HarshaLaxman commented:
Shall I also uprev the version in `setup.py` ?

jacobschaer commented:
You can leave the version string alone - I can roll it and publish when this goes in. Depending on if we can keep the same function interface or not I guess will determine minor/major.

jacobschaer commented:
Thanks for doing this! Some comments, the big thing is hoping we can keep the existing functions as-is since I imagine they're being used and make this it's own thing. If you're not confident in the IPV6 we can leave it IPV4 only until someone who has access to an IPV6 system can test.

@feuerball @MultiName90

jacobschaer commented:
Looks like I need to fix the workflow to exclude python 3.6 now as well :(

HarshaLaxman commented:
> 

- will leave out the option to do IPV6
- will leave the existing `request_vehicle_identification` as is and name this classmethod something else. Any preferences/ideas on the name? `request_vehicle_id` `req_veh_id` `class_request_vehicle_identification` 

And 🥂 for putting this out there! :)

jacobschaer commented:
A quick glance, the spec just calls both use cases "vehicle discovery" and I don't see any particular name for this other than it's the "alt vehicle discovery" scenario lol. 

NI calls it "doip get entities " it looks like. I couldn't find a reference in autosar. So maybe just broadcast_vehicle_discovery or get_entities or similar?

For a future enhancement, these two discovery functions probably should have a version that blocks for a specified amount of time and returns a collection since there could be more than one ECU. But will need some thought and not really important here, just musing.

HarshaLaxman commented:
Renamed to `get_entity` and removed IPV6 and the attached `source_interface` option.

Brought back the original `request_vehicle_identification`

And agree 💯 re the future enhancement!

jacobschaer commented:
Looks good and unit tests pass. Last couple things:

Can you add the new use case to the doc/source/index.rst ? Right now there's an example using the await_vehicle_announcement() workflow. If you could add a new snippet there that would be great for the readthedocs page. 

```
.. code-block:: python

    from doipclient import DoIPClient
    address, announcement = DoIPClient.get_entity()
    logical_address = announcement.logical_address
    ip, port = address
    print(ip, port, logical_address)
```

HarshaLaxman commented:
> 

Added the use case to `index.rst` !

jacobschaer commented:
Merged, thanks for the help
